rust/catalyst-types/src/smt/tree.rs

```diff
@@ -0,0 +1,323 @@
+///
+/// The iterator will yield `2^(ROOT_HEIGHT - target_height)` items.
+///
+/// # Performance Note
+///
+/// Returns an iterator to avoid allocating `2^key_prefix_length` elements upfront.
+/// For heights near the leaves (high values), the theoretical width can be astronomically
+/// large. The iterator allows processing nodes on-demand or early termination.
+fn horizontal_slice<V, H>(
+    store: &DefaultStore<V>,
+    target_height: u8,
+) -> impl Iterator<Item = Result<Option<H256>, Error>>
+where
+    V: Clone,
+    H: sparse_merkle_tree::traits::Hasher + Default,
+{
+    // How long is the path from the root to the target height.
+    #[allow(clippy::arithmetic_side_effects)]
+    let key_prefix_length = ROOT_HEIGHT - target_height;
+
+    // Total number of nodes at this height.
+    let width = 2_u32.pow(u32::from(key_prefix_length));
+
+    // Retrieve all materialized nodes at this height.
+    // SMT nodes are materialized lazily: only nodes on paths to non-empty leaves.
+    // Thus at any height, stored nodes <= theoretical maximum for that level.
+    let materialized_nodes: HashMap<_, _> =
+        materialized_nodes_at_height(store, target_height).collect();
+
+    // For a given height we need to iterate through `width` number of positions,
+    // even if some nodes are not materialized.
+    (0..width).map(move |horizontal_position| {
+        let node_key = utils::node_key(key_prefix_length, horizontal_position)?;
+        let key = BranchKey::new(target_height, node_key);
+
+        Ok(materialized_nodes.get(&node_key).map(|node| {
+            // Recreate the node identity using the provided hasher.
+            let mv = merge::merge::<H>(key.height, &key.node_key, &node.left, &node.right);
+            mv.hash::<H>()
+        }))
+    })
+}
+
+/// Returns an iterator over all materialized nodes at a specific height in the Sparse
+/// Merkle Tree.
+///
+/// # Sparse Merkle Tree Materialization
+///
+/// In a Sparse Merkle Tree (SMT), not all nodes are explicitly stored (materialized) in
+/// memory. Most nodes in the tree are implicitly zero/empty, and only nodes that are on
+/// the path to actual leaf values are materialized and stored. This sparse representation
+/// is what makes SMTs practical for large key spaces - a full binary tree of height 256
+/// would require 2^256 nodes, which is impossible to store.
+///
+/// For example, in a tree with only 3 leaves:
+/// - Height 0 (root): 1 materialized node (the root)
+/// - Height 1: 2 materialized nodes (left and right children of root)
+/// - Height 2: May have 2-3 materialized nodes (only branches leading to leaves)
+/// - Heights 3-255: Sparse - only nodes on paths to the 3 leaves are materialized
+///
+/// This function filters the stored branches to find only those at the requested height,
+/// allowing to examine a horizontal slice of the actually-stored tree structure.
+fn materialized_nodes_at_height<V>(
```

Comment:
This implementation inspects _all branches_ on each call. It might be a performance hit when the tree grows. The solution to this is to implement our own `Store` and then implement `StoreReadOps` and `StoreWriteOps` for it. This way we would be able to keep and update the mapping from height to a `BranchNode` as we modify the tree and retrieve it here at no cost.